cpp/src/gandiva/precompiled/decimal_ops.cc

```diff
@@ -463,5 +465,242 @@ int32_t Compare(const BasicDecimalScalar128& x, const BasicDecimalScalar128& y)
   }
 }
 
+#define DECIMAL_OVERFLOW_IF(condition, overflow) \
+  do {                                           \
+    if (condition) {                             \
+      *overflow = true;                          \
+      return 0;                                  \
+    }                                            \
```

Comment:
we are currently not using the overflow flag in the caller - to be done as part of ARROW-4570. but going forward, I see this to be a single shared flag that is used across an entire batch, involving multiple expressions.

so, flag being set means there was an overflow for atleast one record in atleast one expression. That'll likely be most optimal (rather than examining the flag and saving state for each record in each expression).